### PR TITLE
cmd: implement delta command

### PR DESCRIPTION
**Pull request by btwotch**

### CODE REVIEW

File: cmd/rdiff/delta.go [hunk 1]
```diff
@@ -1,8 +1,63 @@
...
+	"github.com/balena-os/librsync-go"
+	"github.com/sirupsen/logrus"
 	"github.com/urfave/cli"
 )
 
+var (
+	magic            = librsync.BLAKE2_SIG_MAGIC
+	blockLen  uint32 = 512
+	strongLen uint32 = 32
+	bufSize          = 65536
```
Comment (lmbarros) [inline cmd/rdiff/delta.go:16]: We should get these values from the signature file. See my other comment below.

File: cmd/rdiff/delta.go [hunk 2]
```diff
@@ -1,8 +1,63 @@
...
+	}
+	defer delta.Close()
+
+	newfile, err := os.Open(c.Args().Get(2))
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	defer newfile.Close()
+
+	signature, err := librsync.Signature(basis, ioutil.Discard, blockLen, strongLen, magic)
```
Comment (lmbarros) [inline cmd/rdiff/delta.go:53]: As I foreshadowed in comments above, the signature should be passed to the delta command as a command line argument, not computed here.

I know we currently have no means on `master` to read the signature from a file, but I have it implemented on a PR: https://github.com/balena-os/librsync-go/pull/8/files#diff-62f46d82044a129997b958c0f7eb6a95038395d337176418d3dffae98d9332d1R44 (though in this PR the code is not exported, I should probably change that!).

I hope we'll get this PR merged soonish.

File: cmd/rdiff/delta.go [hunk 3]
```diff
@@ -1,8 +1,63 @@
...
+	if c.Args().Get(0) == "" {
+		logrus.Fatalf("Missing basis file")
+	}
+
+	if c.Args().Get(1) == "" {
+		logrus.Fatalf("Missing delta file")
+	}
+	if c.Args().Get(2) == "" {
+		logrus.Fatalf("Missing newfile file")
+	}
```
Comment (lmbarros) [inline cmd/rdiff/delta.go:33]: We should respect the order of command-line arguments that is documented here: https://github.com/balena-os/librsync-go/blob/master/cmd/rdiff/main.go#L45 .

The idea is to keep the output file as the last argument (for consistency with other commands and also to keep the same argument order as in the original `rdiff` tool written in C.)

Also, the first argument should be the signature, not the basis file. More on this on a comment below.

**Review by lmbarros — CHANGES_REQUESTED** [review]
Hi @btwotch, thanks much for taking the time to work on this! My main concern with this PR is that I'd like to keep the behavior of the Go `rdiff` consistent with the behavior of the original (C) `rdiff`.

Please take a look on the review comments, but the gist I believe you'll be able to improve your PR a lot on this regard once we merge https://github.com/balena-os/librsync-go/pull/8 (hopefully soon!).

**Comment by lmbarros** [comment]
@btwotch FYI, we just merged https://github.com/balena-os/librsync-go/pull/8, which includes the [`librsync.ReadSignatureFile()`](https://github.com/balena-os/librsync-go/blob/91ffe9bd011d14760b6e04b2c6eb3c7a70a72ac9/signature.go#L156) function.

File: cmd/rdiff/delta.go (hunk 3 above)
Comment (btwotch) [reply cmd/rdiff/delta.go]: I updated my PR - once tried it with rdiff (from librsync) and with the same parameter with balena/rsync.

File: cmd/rdiff/delta.go (hunk 1 above)
Comment (btwotch) [reply cmd/rdiff/delta.go]: I removed it

File: cmd/rdiff/delta.go (hunk 2 above)
Comment (btwotch) [reply cmd/rdiff/delta.go]: Signature is not calculated anymore in the delta cmd

File: cmd/rdiff/delta.go [hunk 4]
```diff
@@ -1,8 +1,49 @@
...
+
+	delta, err := os.OpenFile(c.Args().Get(2), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	defer newfile.Close()
+
+	err = librsync.Delta(signature, newfile, delta)
+	if err != nil {
+		panic(err)
```
Comment (lmbarros) [inline cmd/rdiff/delta.go:47]:
```suggestion
		logrus.Fatal(err)
```

**Review by lmbarros — APPROVED** [review]
Thanks!

**Pull request merged**